### PR TITLE
Check target voltage plausibility on Transformers and Shunt Compensators

### PR DESCRIPTION
**Pull request by vmouradian**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

Target voltage plausibility is configured via `minPlausibleTargetVoltage`, `maxPlausibleTargetVoltage` and `minNominalVoltageTargetVoltageCheck` parameters.
Today the check is performed only for LfGenerators (Includes generators, batteries, SVCs, HVDC VSC converters).
No check is made for 1/ transformer ratio tap changers and 2/ shunt compensators.

**What is the new behavior (if this is a feature change)?**
Check is made also for 1/ transformer ratio tap changers and 2/ shunt compensators.
As an extra, now the number of disabled controls is reported (instead of only logged).

**Does this PR introduce a breaking change or deprecate an API?**
- [x] No

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/network/impl/AbstractLfBus.java [hunk 1]
```diff
@@ -331,21 +332,34 @@ void setShuntCompensators(List<ShuntCompensator> shuntCompensators, LfNetworkPar
...
     }
 
+    static boolean checkVoltageRegulation(ShuntCompensator shuntCompensator, LfNetworkParameters parameters) {
+        double nominalV = shuntCompensator.getRegulatingTerminal().getVoltageLevel().getNominalV();
+        double targetV = shuntCompensator.getTargetV();
+        if (!shuntCompensator.isVoltageRegulatorOn()) {
+            return false;
+        }
+        if (!VoltageControl.checkTargetV(targetV / nominalV, nominalV, parameters)) {
+            shuntCompensator.setVoltageRegulatorOn(false);
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/impl/AbstractLfBus.java:356]: don't modify the iIDM Network, remove this line. the regulator is disabled only internally in OLF.

File: src/test/java/com/powsybl/openloadflow/ac/AcLoadFlowShuntTest.java [hunk 2]
```diff
@@ -430,6 +430,19 @@ void testIncrementalVoltageControl() {
         assertEquals(0, shunt.getSectionCount());
     }
 
+    @Test
+    void testIncrementalVoltageControlNotPlausibleTargetV() {
```
Comment (jeandemanged) [inline src/test/java/com/powsybl/openloadflow/ac/AcLoadFlowShuntTest.java:434]: why 'incremental and not just `testNotPlausibleTargetV()` ?

this feature is unrelated to incremental control

File: src/main/java/com/powsybl/openloadflow/network/impl/AbstractLfBus.java [hunk 3]
```diff
@@ -331,21 +332,34 @@ void setShuntCompensators(List<ShuntCompensator> shuntCompensators, LfNetworkPar
         if (!parameters.isShuntVoltageControl() && !shuntCompensators.isEmpty()) {
             shunt = new LfShuntImpl(shuntCompensators, network, this, false, parameters, topoConfig);
         } else {
-            List<ShuntCompensator> controllerShuntCompensators = shuntCompensators.stream()
-                    .filter(ShuntCompensator::isVoltageRegulatorOn)
-                    .toList();
+            Map<Boolean, List<ShuntCompensator>> shuntsByVoltageControlCapability = shuntCompensators.stream()
+                    .collect(Collectors.partitioningBy(shunt -> checkVoltageRegulation(shunt, parameters)));
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/impl/AbstractLfBus.java:336]: I kinda preferred the previous way of building the two lists

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java [hunk 4]
```diff
@@ -621,6 +621,12 @@ private static void createTransformerVoltageControl(LfNetwork lfNetwork, RatioTa
         double targetValue = rtc.getTargetV() / regulatingTerminalNominalV;
         Double targetDeadband = rtc.getTargetDeadband() > 0 ? rtc.getTargetDeadband() / regulatingTerminalNominalV : null;
 
+        if (!VoltageControl.checkTargetV(targetValue, controlledBus.getNominalV(), parameters)) {
+            LOGGER.warn("RatioTapChanger on transformer '{}' has an inconsistent target voltage: {} pu: incremental voltage control discarded", controllerBranchId, targetValue);
+            rtc.setRegulating(false);
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:626]: no, don't modify the iIDM network. the regulator is disabled only internally in OLF.

**Review by jeandemanged — CHANGES_REQUESTED** [review]
also:
- LfNetworkLoadingReport to be enhanced with 2 new inconsistencies counters for RTCs and Shunts, similar to generatorsWithInconsistentTargetVoltage
- warning log (see LfNetworkLoaderImpl#create)
- maybe add reports for the 3 inconsistencies (generator, rtc, shunt) ?

**Comment by jeandemanged** [comment]
@vmouradian let's also do some cleanup and remove `linesWithDifferentNominalVoltageAtBothEnds` from `LfNetworkLoadingReport`, this isn't used anymore (should have been removed in #658 but was missed)

File: src/main/java/com/powsybl/openloadflow/network/impl/AbstractLfBus.java [hunk 5]
```diff
@@ -327,25 +327,45 @@ void addBattery(Battery generator, LfNetworkParameters parameters, LfNetworkLoad
...
-            List<ShuntCompensator> fixedShuntCompensators = shuntCompensators.stream()
-                    .filter(sc -> !sc.isVoltageRegulatorOn())
-                    .toList();
             if (!fixedShuntCompensators.isEmpty()) {
                 shunt = new LfShuntImpl(fixedShuntCompensators, network, this, false, parameters, topoConfig);
             }
         }
     }
 
+    static boolean checkVoltageRegulation(ShuntCompensator shuntCompensator, LfNetworkParameters parameters, LfNetworkLoadingReport report) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/AbstractLfBus.java:353]: "Regulation" is never used in OLF, please use "Control" -> checkVoltageControl

Comment (vmouradian) [reply src/main/java/com/powsybl/openloadflow/network/impl/AbstractLfBus.java]: Bad naming indeed, fixed in [b43b026](https://github.com/powsybl/powsybl-open-loadflow/pull/1030/commits/b43b026393ec7b3c609c71f723570534323e320f)

**Review by jeandemanged — CHANGES_REQUESTED** [review]
ok for main code so far for me
in test code, please keep existing tests unchanged, and define 3 new tests for the 3 checks, with 3 test reports txt.

**Comment by vmouradian** [comment]
> ok for main code so far for me in test code, please keep existing tests unchanged, and define 3 new tests for the 3 checks, with 3 test reports txt.

Adressed in [4026142](https://github.com/powsybl/powsybl-open-loadflow/pull/1030/commits/402614283a5716ea6fdc62b40efcfa2326858963)

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java [hunk 6]
```diff
@@ -843,15 +856,12 @@ private LfNetwork create(int numCC, int numSC, Network network, List<Bus> buses,
             LOGGER.warn("Network {}: {} branches have been discarded because connected to same bus at both ends",
                     lfNetwork, report.branchesDiscardedBecauseConnectedToSameBusAtBothEnds);
         }
-        if (report.linesWithDifferentNominalVoltageAtBothEnds > 0) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:846]: why it is removed?

File: src/main/java/com/powsybl/openloadflow/ac/outerloop/SecondaryVoltageControlOuterLoop.java [hunk 7]
```diff
@@ -304,7 +304,7 @@ private Optional<List<String>> processSecondaryVoltageControl(List<LfSecondaryVo
                 .stream()
                 .filter(e -> {
                     double newTargetV = e.getValue();
-                    return LfGenerator.isTargetVoltageNotPlausible(newTargetV, minPlausibleTargetVoltage, maxPlausibleTargetVoltage);
+                    return !VoltageControl.isTargetVoltagePlausible(newTargetV, minPlausibleTargetVoltage, maxPlausibleTargetVoltage);
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/ac/outerloop/SecondaryVoltageControlOuterLoop.java:307]: @annetill just a note, no check on `minNominalVoltageTargetVoltageCheck` here, I have no idea if this is deliberate (I lack knowledge about this outerloop).

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java (hunk 6 above)
Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:846]: this is dead code ... See https://github.com/powsybl/powsybl-open-loadflow/pull/1030#issuecomment-2112800316

Comment (vmouradian) [reply src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:846]: @jeandemanged suggested in a comment above to remove this because it is not used anymore

Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:846]: I think it is a mistake @geofjamg what's happen here?

Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:846]: From here https://github.com/powsybl/powsybl-open-loadflow/pull/658/files#diff-ae21874f7b466e7204256111d8e9f6b3f268bb6d2210efdf6b6c4d30ecf08f92

File: src/main/java/com/powsybl/openloadflow/ac/outerloop/SecondaryVoltageControlOuterLoop.java (hunk 7 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/ac/outerloop/SecondaryVoltageControlOuterLoop.java:307]: I think we are going to wait for the issue @geofjamg . At RTE, we are using this outer loop on network without any generator connected to low nominal voltages, but it could arrived soon.

**Review by annetill — APPROVED** [review]
Thanks!

**Review by jeandemanged — APPROVED** [review]
many thanks @vmouradian :)
